movici_simulation_core/models/rules/expression.py

```diff
@@ -0,0 +1,348 @@
+
+        if len(operands) == 1:
+            return operands[0]
+
+        return BooleanExpression(operator=operator, operands=operands)
+
+
+def _build_grammar() -> pp.ParserElement:
+    """Build the pyparsing grammar for condition expressions.
+
+    :returns: Pyparsing grammar for parsing condition expressions
+    :rtype: pp.ParserElement
+    """
+    integer = pp.Word(pp.nums).setParseAction(lambda t: int(t[0]))
+    decimal = pp.Combine(
+        pp.Optional(pp.Word(pp.nums)) + "." + pp.Word(pp.nums)
+    ).setParseAction(lambda t: float(t[0]))
+    number = decimal | integer
+
+    # Time expressions: durations (34h, 5m, 1d5h30s) and clock times (12:30)
+    time_unit = pp.oneOf("h m s d")
+    single_duration = pp.Combine(pp.Word(pp.nums) + time_unit)
+    duration = pp.Combine(pp.OneOrMore(single_duration)).setParseAction(
+        lambda t: parse_time_value(t[0])
+    )
+    clock_time = pp.Combine(
+        pp.Word(pp.nums) + ":" + pp.Word(pp.nums)
+    ).setParseAction(lambda t: parse_time_value(t[0]))
+
+    boolean = (
+        pp.CaselessKeyword("true").setParseAction(lambda: True)
+        | pp.CaselessKeyword("false").setParseAction(lambda: False)
+    )
+
+    value = duration | clock_time | number | boolean
+
+    comparison_op = pp.oneOf("== != <= >= < >")
+
+    simtime_var = (
+        pp.Literal("<simtime>") | pp.Literal("<simtime") + pp.Literal(">")
+    ).setParseAction(lambda: ExpressionType.SIMTIME)
+
+    clocktime_var = (
+        pp.Literal("<clocktime>")
+        | pp.Literal("<clocktime") + pp.Literal(">")
+        | pp.CaselessKeyword("clocktime")
+    ).setParseAction(lambda: ExpressionType.CLOCKTIME)
+
+    attribute_name = pp.Combine(
+        pp.Word(pp.alphas + "_", pp.alphanums + "_")
+        + pp.Optional("." + pp.Word(pp.alphas + "_", pp.alphanums + "_"))
+    )
+
+    lhs = simtime_var | clocktime_var | attribute_name
+
+    def make_comparison(tokens: pp.ParseResults) -> Comparison:
+        lhs_val = tokens[0]
+        op_str = tokens[1]
+        rhs_val = tokens[2]
+
+        if isinstance(lhs_val, ExpressionType):
+            expr_type = lhs_val
+            attr_name = None
+        else:
+            expr_type = ExpressionType.ATTRIBUTE
+            attr_name = lhs_val
+
+        op_map = {
+            "==": ComparisonOperator.EQ,
+            "!=": ComparisonOperator.NE,
+            "<": ComparisonOperator.LT,
+            "<=": ComparisonOperator.LE,
+            ">": ComparisonOperator.GT,
+            ">=": ComparisonOperator.GE,
+        }
+
+        return Comparison(
+            expr_type=expr_type,
+            operator=op_map[op_str],
+            value=rhs_val,
+            attribute_name=attr_name,
+        )
+
+    comparison = (lhs + comparison_op + value).setParseAction(make_comparison)
+
+    NOT = pp.CaselessKeyword("NOT") | pp.Literal("!")
+    AND = pp.CaselessKeyword("AND") | pp.Literal("&")
+    OR = pp.CaselessKeyword("OR") | pp.Literal("|")
```

Comment:
We're doing boolean comparison here, I think that uses double symbols formally (ie. &&, ||)